src/treesitter.rs

```diff
@@ -21,3 +30,305 @@ impl From<&'_ Node<'_>> for Match {
+}
+
+impl<'a> Iterator for RopeOrSliceRopeTextProviderIterator<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.with_chunks_iterator_mut(|chunks_iterator| chunks_iterator.next())
+    }
+}
+
+// I believe this type can't be Copy/Clone in order for the
+// `get_captures()` unsafe stuff to be sound
+pub struct CaptureInfo<'a> {
+    pub node: Node<'a>,
+    pub pattern_index: usize,
+}
+
+#[self_referencing]
+pub struct Captures<'a, 'text: 'a, 'tree: 'a> {
+    text: RopeOrSlice<'text>,
+    query_cursor: QueryCursor,
+    query: &'a Query,
+    filter: Option<&'a Filterer>,
+    tree: Cow<'tree, Tree>,
+    capture_index: u32,
+    #[borrows(text, mut query_cursor, query, tree)]
+    #[covariant]
+    captures_iterator: QueryCaptures<'this, 'this, 'this, RopeOrSlice<'this>>,
+    #[borrows(tree)]
+    #[covariant]
+    next_capture: Option<CaptureInfo<'this>>,
+}
+
+pub fn get_captures<'a, 'text, 'tree>(
+    language: Language,
+    // text: impl TextProvider<'a> + Parseable,
+    text: impl Into<RopeOrSlice<'text>>,
+    query: &'a Query,
+    capture_index: u32,
+    filter: Option<&'a Filterer>,
+    tree: Option<&'tree Tree>,
+) -> Captures<'a, 'text, 'tree> {
+    let text = text.into();
+    let query_cursor = QueryCursor::new();
+    let tree: Cow<'tree, Tree> = tree.map_or_else(
+        || Cow::Owned(text.parse(&mut get_parser(language), None).unwrap()),
+        Cow::Borrowed,
+    );
+    Captures::new(
+        text,
+        query_cursor,
+        query,
+        filter,
+        tree,
+        capture_index,
+        |text, query_cursor, query, tree| query_cursor.captures(query, tree.root_node(), *text),
+        |_| None,
+    )
+}
+
+impl<'a, 'text, 'tree> StreamingIterator for Captures<'a, 'text, 'tree> {
+    type Item = CaptureInfo<'tree>;
+
+    fn advance(&mut self) {
+        self.with_mut(|all_fields| {
+            for (match_, index_into_query_match_captures) in all_fields.captures_iterator.by_ref() {
+                let this_capture = &match_.captures[index_into_query_match_captures];
+                if this_capture.index != *all_fields.capture_index {
+                    continue;
+                }
+                let single_captured_node = this_capture.node;
+                if all_fields
+                    .filter
+                    .as_ref()
+                    .map_or(true, |filter| filter.call(&single_captured_node))
+                {
+                    *all_fields.next_capture = Some(CaptureInfo {
+                        node: single_captured_node,
+                        pattern_index: match_.pattern_index,
+                    });
+                    return;
+                }
+            }
+            *all_fields.next_capture = None;
+        });
+    }
+
+    fn get<'this>(&'this self) -> Option<&'this Self::Item> {
+        let next_capture = self.borrow_next_capture();
+        // SAFETY: I think this is ok as long as CaptureInfo isn't
```

Comment:
Ok this is where I resorted to `unsafe`

I think I sort of understand why it was flagging this as an error without it

Basically I've declared the associated `Item` type here to be `CaptureInfo<'tree>` but it thinks that what I'd be trying to return is a `CaptureInfo<'this>` (ie its lifetime is dependent on the struct itself aka self-referential), not a `CaptureInfo<'tree>` (which would be allowed to outlive the struct)

And I think it might be right about that but not totally clear on that

But there's no way for me to specify `'this` in the declaration of the `Item` associated type afaik

But it is by definition making the `&` in the return type here a "self-referential lifetime" so what I convinced myself of is that the only way you could "break out of that" and treat the `CaptureInfo` contents as though they're `'tree` when in fact they're `'this` is if you could copy/clone it, and `CaptureInfo` doesn't implement `Copy`/`Clone`

(I'm not even sure that would be a way to achieve anything bad - `CaptureInfo` just has a `usize` and a `Node<'tree>` (which is "actually" `'tree`, I don't see how it couldn't be) on it so I don't think there are like any fields that eg when copied could incorrectly outlive the `&'this` reference - you can safely copy the `Node<'tree>` field)